Feature/lit 2382 js sdk refactor crypto and other wasm based modules

joshLong145 opened the pull request:
# Description

Migrates

- https://github.com/LIT-Protocol/lit-bls-wasm
- https://github.com/LIT-Protocol/lit-ecdsa-wasm-combine
- https://github.com/LIT-Protocol/sev-snp-utils-wasm

To `packages/wasm` which is now being built with each run of our build pipelines and emitting type declarations and new compiled binaries embedded into the binding wrapper. 

- Current size of binding + wasm `393kb` which is approx `1/2` the size of all three modules combined.

- Has upgraded cryptographic implementations with a particular note on now using the `hd-keys-curves` library over hand migrated implementations for `hd key derivation`

- Migrates `packages/crypto` to async implementations with scoped invocation  of `loadModule` to lazy load the `wasm module`

- Updates Github Action Jobs to use rust toolchains for building `packages/wasm`

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

Tested using `crypto.spec.ts` and `local-tests` to confirm that we are compatible with current cryptographic implementations in our nodes. 

Also adds tests to `packages/wasm` for testing of our wasm bindings directly.


# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/LIT-Protocol/js-sdk?pullRequest=503) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/LIT-Protocol/js-sdk?pullRequest=503) before we can accept your contribution.<br/>**2** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: joshLong145<br/>:white_check_mark: mikelodder7<br/>:x: cairomassimo<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/LIT-Protocol/js-sdk?pullRequest=503) it.</sub>

MaximusHaximus commented on packages/core/src/lib/lit-core.ts:
```diff
@@ -477,6 +467,9 @@ export class LitCore {
   }
 
   private async _connect() {
+    if (!this.ready) {
```
Can we get rid of this entirely now that the methods call `initWasm()` internally when they need to?
Can we update `checkSevSnpAttestation()` to call the new wrapper methods that init it for us instead?

joshLong145 replied:
see above comment. if we change the `crypto` apis we will be making breaking changes.

MaximusHaximus reviewed (changes requested):
Holy crap, this is an epic improvement. 💖  I added a few things in-line, but this is looking totally rad 🎸 ✨  :)

joshLong145 commented on packages/wasm/src/index.ts:
```diff
@@ -0,0 +1,150 @@
+ * Allows for lazy loadnig of WASM.
+ */
+
+export async function loadModuleAndBlsCombine(
```
So we currently have to export the methods from the binding because `crypto` does not have async method signatures. I dont think we should change the interfaces in `crypto` at this time as it will force a major version upgrade. so this is why I add `loadModuleAnd` and kept the wrappers in the exports

joshLong145 commented on packages/wasm/src/index.ts:
```diff
@@ -0,0 +1,150 @@
+// loaded wasm module
+let wasmSdkInstance: InitOutput | undefined;
+
+export const initWASM = (): Promise<void> => {
```
See below comment

joshLong145 commented on packages/wasm/src/lib/bls.spec.ts:
```diff
@@ -0,0 +1,48 @@
+// TODO(cairomassimo): move the data into a separate file
```
lol good catch

joshLong145 commented on packages/wasm/tsconfig.base.json:
```diff
@@ -0,0 +1,23 @@
+    "resolveJsonModule": true,
+    "baseUrl": ".",
+    "paths": {
+      "@lit-protocol/wasm-internal": ["./rust/pkg/*"]
```
yes this can be removed.

joshLong145 commented on tools/scripts/tools.mjs:
```diff
@@ -491,6 +491,59 @@ async function buildFunc() {
+    exit();
+  }
+
+  if (BUILD_TYPE === '--packages') {
+    const MODE = args[2];
+    // console.log('MODE:', MODE);
+
+    if (!MODE || MODE === '' || MODE === '--help') {
+      greenLog(
+        `
+            Usage: node tools/scripts/tools.mjs --prebuild --packages [option]
+
+                [option]: the option to run
+            `,
+        true
+      );
+    }
+
+    const ignoreList = (await listDirsRecursive('./apps', false))
+      .map((item) => item.replace('apps/', ''))
+      .join(',');
+
+    const command = `yarn nx run-many --target=prebuild --exclude=${ignoreList}`;
+
+    spawnListener(command, {
+      onDone: () => {
+        console.log('Done!');
+
+        // spawnListener(command, {
+        //   onDone: async () => {
+        //     console.log("Done!");
+        //     await runCommand('yarn postBuild:mapDistFolderNameToPackageJson');
+        exit();
+        //   }
+        // })
+      },
+    });
+  }
+}
+
```
ah yes good catch.

MaximusHaximus reviewed (changes requested):
Just one small tweak to the async loading logic to use a Promise to track loading so concurrent requests can wait on the loading operation instead of returning immediately if loading is already set.  I added inline -- otherwise I think this is GTG

Ansonhkg commented:
Need to add this package to `typedoc.json`

Ansonhkg reviewed (approved):
I tested [here](https://github.com/LIT-Protocol/js-sdk/pull/528) it works great! I think is GTG!

joshLong145 commented:
> Just one small tweak to the async loading logic to use a Promise to track loading so concurrent requests can wait on the loading operation instead of returning immediately if loading is already set. I added inline -- otherwise I think this is GTG

@MaximusHaximus 
I refactored with the changes suggested and it was tested against current `master` which was merged into `staging//v7` and merged back into this branch. So I think we are good to merge unless you see some issues?

MaximusHaximus commented on packages/wasm/src/index.ts:
```diff
@@ -0,0 +1,232 @@
+
+// Give us a promise that _just_ encapsulates initializing the modules so we can wrap it in other logic
+async function initWasm() {
+  await initSync(getModule());
```
```suggestion
  return initSync(getModule());
```

I think we need to return the SDK instance here so that we set `wasmSdkInstance` to the instance on L44 below?

MaximusHaximus commented on packages/wasm/src/index.ts:
```diff
@@ -0,0 +1,232 @@
+    loadingPromise = null;
+  }
+
+  // If we got here, the SDK loaded successfully; just return the instance to the caller
```
```suggestion
  // If we got here, the SDK loaded successfully
```

MaximusHaximus reviewed (changes requested):
@joshLong145 I think this is just about GTG -- I noticed one thing and a small comment change :+1: Otherwise I think this is GTG

joshLong145 commented:
> @joshLong145 I think this is just about GTG -- I noticed one thing and a small comment change 👍 Otherwise I think this is GTG

Awesome, made those updates and looks like checks are all passing

MaximusHaximus reviewed (approved):
<3 awesome work man.  Our first code push landing into the v7 branch :) Good stuff man.
